Reason: actually useful manpages.
Upstream: yes, in master.

The patch below is a heavily trimmed version of it though.
Just user.cfg.in and run_doxygen remains.

From 91315195faa54b42e77d4b62b891ec3a0842f16b Mon Sep 17 00:00:00 2001
From: redi <redi@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Thu, 10 Dec 2015 14:02:52 +0000
Subject: [PATCH] Improve generated libstdc++ API docs

	* doc/doxygen/user.cfg.in: Use EXTENSION_MAPPING tag. Add new headers
	to INPUT. Remove obsolete XML_SCHEMA and XML_DTD tags. Update
	PREDEFINED macros. Set BRIEF_MEMBER_DESC for man-pages.
	* include/backward/strstream: Correct @file comment.
	* include/bits/forward_list.h: Improve Doxygen comments.
	* include/bits/locale_facets_nonio.h: Likewise.
	* include/bits/mutex.h: Likewise.
	* include/bits/shared_ptr.h: Likewise.
	* include/bits/stl_deque.h: Likewise.
	* include/debug/vector (_Safe_vector): Add @brief section to comment.
	* include/experimental/bits/fs_dir.h: Correct @file comment.
	* include/experimental/bits/fs_fwd.h: Likewise.
	* include/experimental/bits/fs_ops.h: Likewise.
	* include/experimental/bits/fs_path.h: Likewise.
	* include/experimental/bits/string_view.tcc: Likewise.
	* include/experimental/optional: Document experimental status.
	* include/experimental/string_view: Correct @file comment.
	* include/ext/pb_ds/detail/bin_search_tree_/traits.hpp: Reduce
	whitespace to avoid Doxygen bug.
	* include/std/bitset: Remove redundant @class Doxygen command. Add
	parentheses to avoid Doxygen bug.
	* include/std/mutex: Improve Doxygen comments.
	* include/tr2/dynamic_bitset: Add missing @param documentation.
	* scripts/run_doxygen: Rename man pages for std::experimental types.

git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/trunk@231512 138bc75d-0d04-0410-961f-82ee72b054a4
---
 libstdc++-v3/ChangeLog                             | 27 +++++++++++
 libstdc++-v3/doc/doxygen/user.cfg.in               | 54 +++++++++++++++-------
 libstdc++-v3/include/backward/strstream            |  6 +--
 libstdc++-v3/include/bits/forward_list.h           |  5 +-
 libstdc++-v3/include/bits/locale_facets_nonio.h    |  2 +-
 libstdc++-v3/include/bits/mutex.h                  | 23 +++++++--
 libstdc++-v3/include/bits/shared_ptr.h             |  2 +-
 libstdc++-v3/include/bits/stl_deque.h              |  1 +
 libstdc++-v3/include/debug/vector                  |  9 ++--
 libstdc++-v3/include/experimental/bits/fs_dir.h    |  2 +-
 libstdc++-v3/include/experimental/bits/fs_fwd.h    |  7 +--
 libstdc++-v3/include/experimental/bits/fs_ops.h    |  5 +-
 libstdc++-v3/include/experimental/bits/fs_path.h   |  2 +-
 .../include/experimental/bits/string_view.tcc      |  2 +-
 libstdc++-v3/include/experimental/optional         |  6 +++
 libstdc++-v3/include/experimental/string_view      |  2 +-
 .../ext/pb_ds/detail/bin_search_tree_/traits.hpp   |  9 +---
 libstdc++-v3/include/std/bitset                    |  4 +-
 libstdc++-v3/include/std/mutex                     |  4 +-
 libstdc++-v3/include/tr2/dynamic_bitset            |  3 ++
 libstdc++-v3/scripts/run_doxygen                   | 17 +++++++
 21 files changed, 140 insertions(+), 52 deletions(-)

diff --git a/libstdc++-v3/doc/doxygen/user.cfg.in b/libstdc++-v3/doc/doxygen/user.cfg.in
index ff2db48..a62eeff 100644
--- a/libstdc++-v3/doc/doxygen/user.cfg.in
+++ b/libstdc++-v3/doc/doxygen/user.cfg.in
@@ -90,7 +90,7 @@ OUTPUT_LANGUAGE        = English
 # documentation (similar to Javadoc). Set to NO to disable this.
 # The default value is: YES.
 
-BRIEF_MEMBER_DESC      = NO
+BRIEF_MEMBER_DESC      = @do_man@
 
 # If the REPEAT_BRIEF tag is set to YES doxygen will prepend the brief
 # description of a member or function before the detailed description
@@ -272,7 +272,7 @@ OPTIMIZE_OUTPUT_VHDL   = NO
 # Note that for custom extensions you also need to set FILE_PATTERNS otherwise
 # the files are not read by doxygen.
 
-EXTENSION_MAPPING      =
+EXTENSION_MAPPING      = no_extension=C++ .h=C++ .tcc=C++ .hpp=C++
 
 # If the MARKDOWN_SUPPORT tag is enabled then doxygen pre-processes all comments
 # according to the Markdown format, which allows for more readable
@@ -757,6 +757,7 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/bitset \
                          include/chrono \
                          include/complex \
+                         include/codecvt \
                          include/condition_variable \
                          include/deque \
                          include/forward_list \
@@ -812,6 +813,7 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/cmath \
                          include/csetjmp \
                          include/csignal \
+                         include/cstdalign \
                          include/cstdarg \
                          include/cstdbool \
                          include/cstddef \
@@ -822,6 +824,7 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/ctgmath \
                          include/ctime \
                          include/cwchar \
+                         include/cuchar \
                          include/cwctype \
                          include/ \
                          include/bits \
@@ -831,6 +834,7 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/backward/hash_set \
                          include/backward/strstream \
                          include/debug \
+                         include/debug/array \
                          include/debug/bitset \
                          include/debug/deque \
                          include/debug/forward_list \
@@ -853,6 +857,7 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/profile/unordered_set \
                          include/profile/vector \
                          include/ext/algorithm \
+                         include/ext/cmath \
                          include/ext/functional \
                          include/ext/iterator \
                          include/ext/memory \
@@ -886,9 +891,36 @@ INPUT                  = @srcdir@/doc/doxygen/doxygroups.cc \
                          include/tr2/ratio \
                          include/tr2/type_traits \
                          include/decimal/decimal \
+                         include/experimental/bits \
+                         include/experimental/algorithm \
                          include/experimental/any \
+                         include/experimental/array \
+                         include/experimental/chrono \
+                         include/experimental/deque \
+                         include/experimental/filesystem \
+                         include/experimental/forward_list \
+                         include/experimental/functional \
+                         include/experimental/iterator \
+                         include/experimental/list \
+                         include/experimental/map \
+                         include/experimental/memory \
+                         include/experimental/memory_resource \
+                         include/experimental/numeric \
                          include/experimental/optional \
+                         include/experimental/propagate_const \
+                         include/experimental/random \
+                         include/experimental/ratio \
+                         include/experimental/regex \
+                         include/experimental/set \
+                         include/experimental/string \
                          include/experimental/string_view \
+                         include/experimental/system_error \
+                         include/experimental/tuple \
+                         include/experimental/type_traits \
+                         include/experimental/unordered_map \
+                         include/experimental/unordered_set \
+                         include/experimental/utility \
+                         include/experimental/vector \
                          include/ext \
                          include/ext/pb_ds \
                          include/ext/pb_ds/detail \
@@ -1965,18 +1997,6 @@ GENERATE_XML           = @do_xml@
 
 XML_OUTPUT             = xml
 
-# The XML_SCHEMA tag can be used to specify a XML schema, which can be used by a
-# validating XML parser to check the syntax of the XML files.
-# This tag requires that the tag GENERATE_XML is set to YES.
-
-XML_SCHEMA             =
-
-# The XML_DTD tag can be used to specify a XML DTD, which can be used by a
-# validating XML parser to check the syntax of the XML files.
-# This tag requires that the tag GENERATE_XML is set to YES.
-
-XML_DTD                =
-
 # If the XML_PROGRAMLISTING tag is set to YES doxygen will dump the program
 # listings (including syntax highlighting and cross-referencing information) to
 # the XML output. Note that enabling this will significantly increase the size
@@ -2111,9 +2131,10 @@ INCLUDE_FILE_PATTERNS  =
 # recursively expanded use the := operator instead of the = operator.
 # This tag requires that the tag ENABLE_PREPROCESSING is set to YES.
 
-PREDEFINED             = __cplusplus=201103L \
+PREDEFINED             = __cplusplus=201402L \
                          __GTHREADS \
                          _GLIBCXX_HAS_GTHREADS \
+                         _GLIBCXX_HAVE_TLS \
                          _GLIBCXX_INCLUDE_AS_CXX11 \
                          "_GLIBCXX_PURE= " \
                          "_GLIBCXX_CONST= " \
@@ -2135,6 +2156,7 @@ PREDEFINED             = __cplusplus=201103L \
                          _GLIBCXX_USE_CONSTEXPR=constexpr \
                          "_GLIBCXX_THROW(E)= " \
                          _GLIBCXX_NOEXCEPT=noexcept \
+                         "_GLIBCXX_NOEXCEPT_IF(C)=noexcept(/*conditional*/)" \
                          _GLIBCXX_NOTHROW=noexcept \
                          _GLIBCXX_USE_NOEXCEPT=noexcept \
                          _GLIBCXX_USE_WCHAR_T \
@@ -2144,7 +2166,7 @@ PREDEFINED             = __cplusplus=201103L \
                          _GLIBCXX_USE_NANOSLEEP \
                          __cpp_exceptions \
                          __cpp_rtti \
-                         ATOMIC_INT_LOCK_FREE \
+                         ATOMIC_INT_LOCK_FREE=2 \
                          PB_DS_DATA_TRUE_INDICATOR \
                          PB_DS_STATIC_ASSERT=// \
                          "_GLIBCXX_BEGIN_NAMESPACE_ALGO= " \
diff --git a/libstdc++-v3/scripts/run_doxygen b/libstdc++-v3/scripts/run_doxygen
index 461adaa..021ebd0 100644
--- a/libstdc++-v3/scripts/run_doxygen
+++ b/libstdc++-v3/scripts/run_doxygen
@@ -334,6 +334,23 @@ for f in *__profile_*; do
     mv $f $newname
 done
 
+# Remove inline namespaces used for versioning.
+for f in *_V2_*; do
+    newname=`echo $f | sed 's/_V2_/::/'`
+    sed 's/::_V2::/::/g' $f > $newname
+    rm $f
+done
+#for f in *_experimental_filesystem_v?_*; do
+#    newname=`echo $f | sed 's/_filesystem_v._/::filesystem::/'`
+#    sed 's/::filesystem::v.::/::filesystem::/g' $f > $newname
+#    rm $f
+#done
+for f in *experimental_fundamentals_v?_*; do
+    newname=`echo $f | sed 's/experimental_.*_v[[:digit:]]_/experimental::/'`
+    sed 's/::experimental::fundamentals_v[[:digit:]]::/::experimental::/g' $f > $newname
+    rm $f
+done
+
 # Then, clean up other top-level namespaces.
 for f in std_tr1_*; do
     newname=`echo $f | sed 's/^std_tr1_/std::tr1::/'`
